src/GraphQL.MicrosoftDI/GraphQLBuilder.cs

```diff
@@ -0,0 +1,126 @@
+#nullable enable
+
+using System;
+using GraphQL.DI;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.DependencyInjection.Extensions;
+using Microsoft.Extensions.Options;
+using MSServiceLifetime = Microsoft.Extensions.DependencyInjection.ServiceLifetime;
+using ServiceLifetime = GraphQL.DI.ServiceLifetime;
+
+namespace GraphQL.MicrosoftDI
+{
+    /// <summary>
+    /// An implementation of <see cref="IGraphQLBuilder"/> which uses the Microsoft dependency injection framework
+    /// to register services and configure options.
+    /// </summary>
+    internal class GraphQLBuilder : GraphQLBuilderBase
+    {
+        private readonly IServiceCollection _services;
+        /// <summary>
+        /// Initializes a new instance for the specified service collection.
+        /// </summary>
+        /// <remarks>
+        /// Registers various default services via <see cref="GraphQLBuilderBase.Initialize"/>.
+        /// </remarks>
+        public GraphQLBuilder(IServiceCollection services)
+        {
+            _services = services ?? throw new ArgumentNullException(nameof(services));
+            services.AddOptions();
+            Initialize();
+        }
+
+        private static MSServiceLifetime TranslateLifetime(ServiceLifetime serviceLifetime)
+            => serviceLifetime switch
+            {
+                ServiceLifetime.Singleton => MSServiceLifetime.Singleton,
+                ServiceLifetime.Scoped => MSServiceLifetime.Scoped,
+                ServiceLifetime.Transient => MSServiceLifetime.Transient,
+                _ => throw new ArgumentOutOfRangeException(nameof(serviceLifetime))
+            };
+
+        /// <inheritdoc/>
+        public override IGraphQLBuilder Configure<TOptions>(Action<TOptions, IServiceProvider>? action = null)
+        {
+            this.TryRegister(services => services.GetService<IOptions<TOptions>>()?.Value ?? new TOptions(), DI.ServiceLifetime.Singleton);
+            if (action != null)
+            {
+                this.Register<IConfigureOptions<TOptions>>(services => new ConfigureNamedOptions<TOptions>(Options.DefaultName, opt => action(opt, services)), ServiceLifetime.Singleton);
```

Comment:
Worth to add comment
> This is used instead of "normal" services.Configure(configureOptions) to pass IServiceProvider to user code.

(from server project)

Reply:
#2781